cfme/markers/env_markers/provider.py

```diff
@@ -14,6 +19,59 @@
 ONE_PER_TYPE = 'one_per_type'
 
 
+class DPFilter(ProviderFilter):
+    def __call__(self, provider):
+        """ Applies this filter on a given DataProvider
+
+        Usage:
+            pf = ProviderFilter('cloud_infra', categories=['cloud', 'infra'])
+            providers = list_providers([pf])
+            pf2 = ProviderFilter(
+                classes=[GCEProvider, EC2Provider], required_fields=['small_template'])
+            provider_keys = [prov.key for prov in list_providers([pf, pf2])]
+            ^ this will list keys of all GCE and EC2 providers
+            ...or...
+            pf = ProviderFilter(required_tags=['openstack', 'complete'])
+            pf_inverted = ProviderFilter(required_tags=['disabled'], inverted=True)
+            providers = list_providers([pf, pf_inverted])
+            ^ this will return providers that have both the "openstack" and "complete" tags set
+              and at the same time don't have the "disabled" tag
+            ...or...
+            pf = ProviderFilter(keys=['rhevm34'], class=CloudProvider, conjunctive=False)
+            providers = list_providers([pf])
+            ^ this will list all providers that either have the 'rhevm34' key or are an instance
+              of the CloudProvider class and therefore are a cloud provider
+
+        Returns:
+            `True` if provider passed all checks and was not filtered out, `False` otherwise.
+            The result is opposite if the 'inverted' attribute is set to `True`.
+        """
+        classes_l = self._filter_classes(provider)
+        results = [classes_l]
+        if not provider.one_of(DataProvider):
+            from cfme.common.provider import BaseProvider
+            if provider.one_of(BaseProvider):
```

Comment:
Actually,  both DataProvider and BaseProvider come to this filter at collection stage one after other.
[Here](https://github.com/ManageIQ/integration_tests/pull/7006/files/744e0db76953d5e35b3d6685560bb5860f14b92d#diff-bb9826babc59e31a3994a52635379668R241) this filter is called with one BaseProviders as an argument.
[There](https://github.com/ManageIQ/integration_tests/pull/7006/files/744e0db76953d5e35b3d6685560bb5860f14b92d#diff-bb9826babc59e31a3994a52635379668R245) this filter is called with DataProvider.
And then left providers are tied to each other.
Actually, there is a some mess because we put together both DataProvider filters and Provider filters. DataProvider filters are applied against DataProviders whilst  Provider filters filter out Providers.
We already discussed this with @psav and decided to turn this stuff into two separate stages on next step.